locations/jclouds/src/main/java/brooklyn/location/jclouds/JcloudsLocation.java

```diff
@@ -2664,6 +2664,59 @@ private String getPublicHostnameAws(HostAndPort sshHostAndPort, ConfigBag setup)
         }
     }
 
+    /**
+     * Attempts to obtain the private hostname or IP of the node, as advertised by the cloud provider.
+     * 
+     * For some clouds (e.g. aws-ec2), it will attempt to find the hostname (as that works in public+private).
+     */
+    protected String getPrivateHostname(NodeMetadata node, Optional<HostAndPort> sshHostAndPort, ConfigBag setup) {
+        String provider = (setup != null) ? setup.get(CLOUD_PROVIDER) : null;
+        if (provider == null) provider= getProvider();
+
+        if ("aws-ec2".equals(provider)) {
```

Comment:
This smells. I can live with it as a workaround in this one exceptional case, because the cost of refactoring to introduce a provider-agnostic abstraction is high. Let's not allow the special cases to proliferate though. Please add a comment to discourage others from following this example.